lib/src/ably_implementation.dart

```diff
@@ -24,79 +25,75 @@ class AblyImplementation implements Ably {
   /// Storing all platform objects, for easy references/cleanup
   final List<PlatformObject> _platformObjects = [];
 
-  /// ably handle, created from platform's end
-  /// This makes sure we are accessing the right Ably instance on platform
-  /// as users can create multiple [Ably] instances
-  int _handle;
-
   factory AblyImplementation() {
     /// Uses our custom message codec so that we can pass Ably types to the
     /// platform implementations.
     StandardMethodCodec codec = StandardMethodCodec(Codec());
     return AblyImplementation._constructor(
-        MethodChannel('io.ably.flutter.plugin', codec),
-        StreamsChannel('io.ably.flutter.stream', codec)
+      MethodChannel('io.ably.flutter.plugin', codec),
+      StreamsChannel('io.ably.flutter.stream', codec)
     );
   }
 
+  bool _initialized = false;
+
   AblyImplementation._constructor(this.methodChannel, this.streamsChannel);
 
-  /// Registering instance with ably.
-  /// On registration, older ably instance id destroyed!
-  /// TODO check if this is desired behavior in case if 2 different ably instances are created in app
-  Future<int> _register() async => (null != _handle) ? _handle : _handle = await methodChannel.invokeMethod(PlatformMethod.registerAbly);
+  void _init(){
+    methodChannel.invokeMethod(PlatformMethod.registerAbly).then((_){
+      _initialized = true;
+    });
+  }
+
+  /// Initializing ably on platform side by invoking `register` platform method.
+  /// Register will clear any stale instances on platform.
+  Future _initialize() async {
+    if(_initialized) return;
+    _init();
+    // if `_initialized` is false => initialization is in progress.
+    // Let's wait for 250ms and retry
+    //
+    // this is required as many asynchronous `_initialize` calls
+    // will be invoked from different Ably instances
+    while(true){
+      await Future.delayed(Duration(milliseconds: 250));
```

Comment:
Another seemingly arbitrary delay. Quite a long time, in CPU terms.

Reply:
@QuintinWillison what could be the optimistic wait time? This has to do a round-trip - a platform method call

Reply:
I think my [other response](https://github.com/ably/ably-flutter/pull/18#discussion_r450699980) equally applies here.

Reply:
Fixed @ c792574efd37a3cff749aa066970262d0ff6dfec